src/types.rs

```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
         }
     }
 }
+
+pub struct CastPass;
+
+declare_lint!(pub CAST_PRECISION_LOSS, Allow,
+              "casts that cause loss of precision");
+declare_lint!(pub CAST_SIGN_LOSS, Allow,
+              "casts from signed types to unsigned types");
+declare_lint!(pub CAST_POSSIBLE_OVERFLOW, Allow,
+              "casts that may cause overflow");
+
+impl LintPass for CastPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(CAST_PRECISION_LOSS,
+                    CAST_SIGN_LOSS,
+                    CAST_POSSIBLE_OVERFLOW)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &Expr) {
+        if let ExprCast(ref ex, _) = expr.node {
+            let (cast_from, cast_to) = (cx.tcx.expr_ty(&*ex), cx.tcx.expr_ty(expr));
+            if cast_from.is_numeric() && !in_external_macro(cx, expr.span) {
+                match (cast_from.is_integral(), cast_to.is_integral()) {
+                    (true, false)  => {
+                        match (&cast_from.sty, &cast_to.sty) {
+                            (&ty::TypeVariants::TyInt(i), &ty::TypeVariants::TyFloat(f)) => {
+                                match (i, f) {
+                                    (ast::IntTy::TyI32, ast::FloatTy::TyF32) |
+                                    (ast::IntTy::TyI64, ast::FloatTy::TyF32) |
+                                    (ast::IntTy::TyI64, ast::FloatTy::TyF64) => {
+                                        span_lint(cx, CAST_PRECISION_LOSS, expr.span,
+                                                  &format!("converting from {} to {}, which causes a loss of precision",
```

Comment:
nit: perhaps some more text here explaining why this is a loss of precision? (intX has X bits for the mantissa, floatX has less than X bits for the mantissa)